File: DevExpress VCL/ExpressVerticalGrid/Demos/CBuilder/EditorsLookupDemo/EditorsLookupDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "EditorsLookupDemoMain.h"
#include "EditorsLookupDemoData.h"
#include "EditorsLookupDemoNewUser.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxCalendar"
#pragma link "cxCheckBox"
#pragma link "cxContainer"
#pragma link "cxDBEdit"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDBLookupEdit"
#pragma link "cxDBNavigator"
#pragma link "cxDropDownEdit"
#pragma link "cxImageComboBox"
#pragma link "cxLookupEdit"
#pragma link "cxMaskEdit"
#pragma link "cxMemo"
#pragma link "cxNavigator"
#pragma link "cxPropertiesStore"
#pragma link "cxTextEdit"
#pragma link "cxLookAndFeelPainters"
#pragma resource "*.dfm"
TEditorsLookupDemoMainForm *EditorsLookupDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TEditorsLookupDemoMainForm::TEditorsLookupDemoMainForm(TComponent* Owner)
  : TDemoBasicMainForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TEditorsLookupDemoMainForm::SetDefaultLookAndFeel()
{
//
}

void __fastcall TEditorsLookupDemoMainForm::FormCreate(TObject *Sender)
{
  TDemoBasicMainForm::FormCreate(Sender);
  FStream = new TMemoryStream();
  cxPropertiesStore->StorageStream = FStream;
  miStoringGridLookFeel->Checked = true;
  miStoringGridLookFeelClick(miStoringGridLookFeel);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsLookupDemoMainForm::FormDestroy(TObject *Sender)
{
  if (cxPropertiesStore->StorageType == stStream)
    miIniStoreType->Click();
  delete FStream;
}
//---------------------------------------------------------------------------

void __fastcall TEditorsLookupDemoMainForm::lcbCreatorNewLookupDisplayText(TObject *Sender,
    const TCaption AText)
{
  TcxDBLookupComboBox *ALookupControl = (TcxDBLookupComboBox*)Sender;
  ALookupControl->Reset();
  if (EditorsLookupDemoNewUserForm->ShowEx(AText) == mrOk) {
    ALookupControl->Text = EditorsLookupDemoDataDM->tblUsersUserName->AsString;
    ALookupControl->DataBinding->Field->Value =
      EditorsLookupDemoDataDM->tblUsersID->Value;
  }
  Abort();
}
//---------------------------------------------------------------------------


void __fastcall TEditorsLookupDemoMainForm::ChangeLookupModeClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = true;
  switch (((TMenuItem*)Sender)->Tag){
    case 1: SetPickLookupMode(); break;
    case 2: SetStandardLookupMode(); break;
    case 3: SetEditLookupMode(); break;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsLookupDemoMainForm::StorageTypeClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = true;
  ((TMenuItem*)Sender)->Parent->Tag = ((TMenuItem*)Sender)->Tag;
  cxPropertiesStore->StorageType = (TcxStorageType)((TMenuItem*)Sender)->Tag;
  if (((TMenuItem*)Sender)->Tag == 0)
    cxPropertiesStore->StorageName = cxPropertiesStore->Name + ".ini";
  else
    cxPropertiesStore->StorageName = cxPropertiesStore->Name;
}
//---------------------------------------------------------------------------

void __fastcall TEditorsLookupDemoMainForm::miStoreClick(TObject *Sender)
{
  cxPropertiesStore->StoreTo(true);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsLookupDemoMainForm::miRestoreClick(TObject *Sender)
{
  FStream->Position = 0;
  cxPropertiesStore->RestoreFrom();
}
//---------------------------------------------------------------------------


TMenuItem* __fastcall GetLookAndFeelMenuItem(TMenuItem* AParentMenuItem)
{
  for (int i = 0; i < AParentMenuItem->Count; i++)
	if (AParentMenuItem->Items[i]->Caption == "&Look&&Feel")
	{
	  return(AParentMenuItem->Items[i]);
	};
  return(NULL);
}

void __fastcall TEditorsLookupDemoMainForm::miStoringGridLookFeelClick(TObject *Sender)
{
  int AIndex = IndexOfPropertiesStoreComponent(cxLookAndFeelController);
  TMenuItem* ALookAndFeelMenuItem = GetLookAndFeelMenuItem(miOptions);
  if (((TMenuItem*)Sender)->Checked){
    if (AIndex > -1){
      cxPropertiesStore->Components->ComponentItems[AIndex]->Properties->Add("NativeStyle");
      cxPropertiesStore->Components->ComponentItems[AIndex]->Properties->Add("Kind");
    }
    else{
      TcxPropertiesStoreComponent *AStoreComponent =
        (TcxPropertiesStoreComponent*)cxPropertiesStore->Components->Add();
      AStoreComponent->Component = cxLookAndFeelController;
      AStoreComponent->Properties->Add("NativeStyle");
      AStoreComponent->Properties->Add("Kind");
    }
	for (int i = 0; i < ALookAndFeelMenuItem->Count; i++)
	  {AddMenuItemStoreComponent(ALookAndFeelMenuItem->Items[i]);};
  }
  else
    if (AIndex > -1){
	  delete cxPropertiesStore->Components->Items[AIndex];
	  for (int i = 0; i < ALookAndFeelMenuItem->Count; i++)
	  {
		AIndex = IndexOfPropertiesStoreComponent(ALookAndFeelMenuItem->Items[i]);
        delete cxPropertiesStore->Components->Items[AIndex];
	  };
    }
}

//---------------------------------------------------------------------------
int TEditorsLookupDemoMainForm::IndexOfPropertiesStoreComponent(TComponent *AComponent)
{
  for (int I = 0; I < cxPropertiesStore->Components->Count; I++)
    if (cxPropertiesStore->Components->ComponentItems[I]->Component == AComponent)
      return I;
  return -1;
}

//---------------------------------------------------------------------------
void TEditorsLookupDemoMainForm::AddMenuItemStoreComponent(TComponent *AItem)
{
  int AIndex = IndexOfPropertiesStoreComponent(AItem);
  if (AIndex > -1)
    cxPropertiesStore->Components->ComponentItems[AIndex]->Properties->Add("Checked");
  else{
    TcxPropertiesStoreComponent *AStoreComponent =
      (TcxPropertiesStoreComponent*)cxPropertiesStore->Components->Add();
    AStoreComponent->Component = AItem;
    AStoreComponent->Properties->Add("Checked");
  }
}

//---------------------------------------------------------------------------
void __fastcall TEditorsLookupDemoMainForm::miStorageActiveClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  cxPropertiesStore->Active = ((TMenuItem*)Sender)->Checked;
  miStoringGridLookFeel->Enabled = ((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void TEditorsLookupDemoMainForm::SetPickLookupMode()
{
  lcbCreator->Properties->DropDownListStyle = lsEditFixedList;
  lcbCreator->Properties->ImmediateDropDown = true;
  lcbCreator->Properties->IncrementalFiltering = true;
  lcbOwner->Properties->DropDownListStyle = lsEditFixedList;
  lcbOwner->Properties->ImmediateDropDown = true;
  lcbOwner->Properties->IncrementalFiltering = true;
  lcbProject->Properties->DropDownListStyle = lsEditFixedList;
  lcbProject->Properties->IncrementalFiltering = True;
}
//---------------------------------------------------------------------------

void TEditorsLookupDemoMainForm::SetStandardLookupMode()
{
  lcbCreator->Properties->DropDownListStyle = lsFixedList;
  lcbCreator->Properties->ImmediateDropDown = true;
  lcbCreator->Properties->IncrementalFiltering = false;
  lcbOwner->Properties->DropDownListStyle = lsFixedList;
  lcbOwner->Properties->ImmediateDropDown = true;
  lcbOwner->Properties->IncrementalFiltering = false;
  lcbProject->Properties->DropDownListStyle = lsFixedList;
  lcbProject->Properties->IncrementalFiltering = false;
  lcbProject->Properties->ImmediateDropDown = true;
}
//---------------------------------------------------------------------------

void TEditorsLookupDemoMainForm::SetEditLookupMode()
{
  lcbCreator->Properties->DropDownListStyle = lsEditList;
  lcbCreator->Properties->ImmediateDropDown = false;
  lcbCreator->Properties->IncrementalFiltering = false;
  lcbOwner->Properties->DropDownListStyle = lsEditList;
  lcbOwner->Properties->ImmediateDropDown = false;
  lcbOwner->Properties->IncrementalFiltering = false;
  lcbProject->Properties->DropDownListStyle = lsEditFixedList;
  lcbProject->Properties->IncrementalFiltering = false;
}
//---------------------------------------------------------------------------

File: DevExpress VCL/ExpressVerticalGrid/Demos/Delphi/EditorsLookupDemo/EditorsLookupDemoMain.pas
unit EditorsLookupDemoMain;

interface

uses
  Windows, Messages, Classes, Graphics, Controls, Forms, Dialogs, cxControls,
  cxLookupGrid, cxLookupDBGrid, cxMaskEdit, StdCtrls, ExtCtrls, cxContainer,
  cxEdit, cxTextEdit, cxDropDownEdit, cxDBEdit, cxStyles, Menus, ActnList,
  ImgList, ShellAPI, ComCtrls, cxLookAndFeels, cxLookupEdit, cxDBLookupComboBox,
  DBCtrls, cxCustomData, cxGraphics, cxFilter, cxData, DB, cxDBData, cxClasses,
  cxCalendar, cxImageComboBox, cxMemo, cxCheckBox, cxButtons, cxNavigator,
  cxDBLookupEdit, cxStorage, cxPropertiesStore, cxDBNavigator, DemoBasicMain,
  cxLookAndFeelPainters;

type
  TEditorsLookupDemoMainForm = class(TDemoBasicMainForm)
    miStorage: TMenuItem;
    miStorageType: TMenuItem;
    miStore: TMenuItem;
    miRestore: TMenuItem;
    miIniStoreType: TMenuItem;
    miRegistryStoreType: TMenuItem;
    miMemoryStoreType: TMenuItem;
    N5: TMenuItem;
    N3: TMenuItem;
    miStoringGridLookFeel: TMenuItem;
    N6: TMenuItem;
    miStorageActive: TMenuItem;
    N1: TMenuItem;
    LookupOptions1: TMenuItem;
    miPickMode: TMenuItem;
    miStandardMode: TMenuItem;
    miEditMode: TMenuItem;
    cxPropertiesStore: TcxPropertiesStore;
    cxStorageActiveStore: TcxPropertiesStore;
    pnlEditors: TPanel;
    gbIssue: TGroupBox;
    Label1: TLabel;
    Label2: TLabel;
    Label9: TLabel;
    edName: TcxDBTextEdit;
    meDescription: TcxDBMemo;
    lcbCreator: TcxDBLookupComboBox;
    gbInfo: TGroupBox;
    Label3: TLabel;
    Label4: TLabel;
    cbPriority: TcxDBImageComboBox;
    deCreateDate: TcxDBDateEdit;
    chbRequest: TcxDBCheckBox;
    gbStatus: TGroupBox;
    Label5: TLabel;
    Label6: TLabel;
    Label7: TLabel;
    cbStatus: TcxDBImageComboBox;
    deLastModifiedDate: TcxDBDateEdit;
    deFixedDate: TcxDBDateEdit;
    DBNavigator: TcxDBNavigator;
    gbProject: TGroupBox;
    Label8: TLabel;
    Label10: TLabel;
    lcbProject: TcxDBLookupComboBox;
    lcbOwner: TcxDBLookupComboBox;
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure ChangeLookupModeClick(Sender: TObject);
    procedure lcbCreatorNewLookupDisplayText(Sender: TObject;
      const AText: TCaption);
    procedure miStoreClick(Sender: TObject);
    procedure miRestoreClick(Sender: TObject);
    procedure miStoringGridLookFeelClick(Sender: TObject);
    procedure miStorageActiveClick(Sender: TObject);
    procedure StorageTypeClick(Sender: TObject);
  protected
    procedure SetDefaultLookAndFeel; override;
  private
    FStream: TMemoryStream;
    procedure SetPickLookupMode;
    procedure SetStandardLookupMode;
    procedure SetEditLookupMode;
  end;

var
  EditorsLookupDemoMainForm: TEditorsLookupDemoMainForm;

implementation

uses EditorsLookupDemoData, EditorsLookupDemoNewUser, SysUtils;

{$R *.dfm}

procedure TEditorsLookupDemoMainForm.FormCreate(Sender: TObject);
begin
  inherited;
  FStream := TMemoryStream.Create;
  cxPropertiesStore.StorageStream := FStream;
  miStoringGridLookFeel.Checked := True;
  miStoringGridLookFeelClick(miStoringGridLookFeel);
end;

procedure TEditorsLookupDemoMainForm.FormDestroy(Sender: TObject);
begin
  if cxPropertiesStore.StorageType = stStream then
    miIniStoreType.Click;
  FreeAndNil(FStream);
end;

procedure TEditorsLookupDemoMainForm.lcbCreatorNewLookupDisplayText(
  Sender: TObject; const AText: TCaption);
var
  ALookupControl: TcxDBLookupComboBox;
begin
  ALookupControl := TcxDBLookupComboBox(Sender);
  ALookupControl.Reset;
  if (EditorsLookupDemoNewUserForm.ShowEX(AText) = mrOK) then
  begin
    ALookupControl.Text := EditorsLookupDemoDataDM.tblUsersUserName.AsString;
    ALookupControl.DataBinding.Field.Value :=
      EditorsLookupDemoDataDM.tblUsersID.Value;
  end;
  Abort;
end;

procedure TEditorsLookupDemoMainForm.ChangeLookupModeClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := True;
  case TMenuItem(Sender).Tag of
    1: SetPickLookupMode;
    2: SetStandardLookupMode;
    3: SetEditLookupMode;
  end
end;

procedure TEditorsLookupDemoMainForm.SetPickLookupMode;
begin
  with lcbCreator.Properties do
  begin
    DropDownListStyle := lsEditFixedList;
    ImmediateDropDown := True;
    IncrementalFiltering := True;
  end;
  with lcbOwner.Properties do
  begin
    DropDownListStyle := lsEditFixedList;
    ImmediateDropDown := True;
    IncrementalFiltering := True;
  end;
  with lcbProject.Properties do
  begin
    DropDownListStyle := lsEditFixedList;
    IncrementalFiltering := True;
  end;
end;

procedure TEditorsLookupDemoMainForm.SetStandardLookupMode;
begin
  with lcbCreator.Properties do
  begin
    DropDownListStyle := lsFixedList;
    IncrementalFiltering := False;
    ImmediateDropDown := True;
  end;
  with lcbOwner.Properties do
  begin
    DropDownListStyle := lsFixedList;
    IncrementalFiltering := False;
    ImmediateDropDown := True;
  end;
  with lcbProject.Properties do
  begin
    DropDownListStyle := lsFixedList;
    IncrementalFiltering := False;
    ImmediateDropDown := True;
  end;
end;

procedure TEditorsLookupDemoMainForm.SetEditLookupMode;
begin
  with lcbCreator.Properties do
  begin
    DropDownListStyle := lsEditList;
    IncrementalFiltering := False;
    ImmediateDropDown := False;
  end;
  with lcbOwner.Properties do
  begin
    DropDownListStyle := lsEditList;
    IncrementalFiltering := False;
    ImmediateDropDown := False;
  end;
  with lcbProject.Properties do
  begin
    DropDownListStyle := lsEditFixedList;
    IncrementalFiltering := False;
  end;
end;

procedure TEditorsLookupDemoMainForm.StorageTypeClick(Sender: TObject);
begin
  TMenuItem(Sender).Checked := True;
  TMenuItem(Sender).Parent.Tag := TMenuItem(Sender).Tag;
  cxPropertiesStore.StorageType := TcxStorageType(TMenuItem(Sender).Tag);
  if TMenuItem(Sender).Tag = 0 then
    cxPropertiesStore.StorageName := cxPropertiesStore.Name + '.ini'
  else
    cxPropertiesStore.StorageName := cxPropertiesStore.Name;
end;

procedure TEditorsLookupDemoMainForm.miStoreClick(Sender: TObject);
begin
  cxPropertiesStore.StoreTo;
end;

procedure TEditorsLookupDemoMainForm.miRestoreClick(Sender: TObject);
begin
  FStream.Position := 0;
  cxPropertiesStore.RestoreFrom;
end;

procedure TEditorsLookupDemoMainForm.miStoringGridLookFeelClick(
  Sender: TObject);
var
 AIndex: Integer;

  function IndexOfPropertiesStoreComponent(AComponent: TComponent): Integer;
  begin
    for Result := 0 to cxPropertiesStore.Components.Count - 1 do
      if cxPropertiesStore.Components[Result].Component = AComponent then
        Exit;
    Result := -1;
  end;

  procedure AddMenuItemStoreComponent(AItem: TComponent);
  begin
    AIndex := IndexOfPropertiesStoreComponent(AItem);
    if AIndex > -1 then
      cxPropertiesStore.Components[AIndex].Properties.Add('Checked')
    else
      with TcxPropertiesStoreComponent(cxPropertiesStore.Components.Add) do
      begin
        Component := AItem;
        Properties.Add('Checked');
      end;
  end;

  function GetLookAndFeelMenuItem: TMenuItem;
  var
    I: Integer;
  begin
    Result := nil;
    for I := 0 to miOptions.Count - 1 do
      if miOptions.Items[I].Caption = '&Look&&Feel' then
      begin
        Result := miOptions.Items[I];
        Break;
      end;
  end;

var
  ALookAndFeelMenuItem: TMenuItem;
  I: Integer;  
begin
  AIndex := IndexOfPropertiesStoreComponent(cxLookAndFeelController);
  ALookAndFeelMenuItem := GetLookAndFeelMenuItem;
  if TMenuItem(Sender).Checked then
  begin
    if AIndex > -1 then
    begin
      cxPropertiesStore.Components[AIndex].Properties.Add('NativeStyle');
      cxPropertiesStore.Components[AIndex].Properties.Add('Kind');
    end
    else
      with TcxPropertiesStoreComponent(cxPropertiesStore.Components.Add) do
      begin
        Component := cxLookAndFeelController;
        Properties.Add('NativeStyle');
        Properties.Add('Kind');
      end;
    for I := 0 to ALookAndFeelMenuItem.Count - 1 do
      AddMenuItemStoreComponent(ALookAndFeelMenuItem[I]);
  end
  else
    if AIndex > -1 then
    begin
      cxPropertiesStore.Components[AIndex].Free;
      for I := 0 to ALookAndFeelMenuItem.Count - 1 do
      begin
        AIndex := IndexOfPropertiesStoreComponent(ALookAndFeelMenuItem[I]);
        cxPropertiesStore.Components[AIndex].Free;
      end;
    end;
end;

procedure TEditorsLookupDemoMainForm.miStorageActiveClick(
  Sender: TObject);
begin
  TMenuItem(Sender).Checked := not TMenuItem(Sender).Checked;
  cxPropertiesStore.Active := TMenuItem(Sender).Checked;
  miStoringGridLookFeel.Enabled := TMenuItem(Sender).Checked;
end;

procedure TEditorsLookupDemoMainForm.SetDefaultLookAndFeel;
begin
  //
end;

end.
